server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -58,58 +65,76 @@ public static MailDto fromMail(Mail mail, List<AdditionalFields> additionalField
+        if (!additionalFields.contains(AdditionalField.BODY)) {
             return Optional.empty();
         }
+
         try {
-            return Optional.ofNullable(mail.getMessage().getContent().toString());
-        } catch (IOException | MessagingException | NullPointerException e) {
-            throw new MissingRequestedField("body");
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function((Message message) -> extractor.extract(message)).sneakyThrow())
+                    .flatMap(extractBody());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.BODY);
         }
     }
 
-    private static Optional<Map<String, List<String>>> fetchHeaders(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
-        if (!additionalFields.contains(AdditionalFields.HEADERS)) {
+    private static Function<? super MessageContent, Optional<String>> extractBody() {
+        return (content) -> {
+            Optional<String> body = content.getTextBody();
+            if (body.isPresent()) {
+                return body;
+            } else {
+                return content.getHtmlBody();
+            }
+        };
+    }
+
+    private static Message convertMessage(MimeMessage message) throws IOException, MessagingException {
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        message.writeTo(rawMessage);
+        return org.apache.james.mime4j.dom.Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(new ByteArrayInputStream(rawMessage.toByteArray()))
+                .build();
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
             return Optional.empty();
         }
-        Map<String, List<String>> headers = new HashMap<>();
 
-        Enumeration<Header> rawHeaders;
         try {
-            rawHeaders = mail.getMessage().getAllHeaders();
-            while (rawHeaders.hasMoreElements()) {
-                Header header = rawHeaders.nextElement();
-                String name = header.getName();
-                if (!headers.containsKey(name)) {
-                    headers.put(name, new ArrayList<>());
-                }
-
-                headers.get(name).add(header.getValue());
-            }
+            HeadersDto headers = new HeadersDto();
+            Collections
+                .list(mail.getMessage().getAllHeaders())
+                .forEach((header) -> headers.add(header.getName(), header.getValue()));
```

Comment:
HeadersDTO should not be subject to side effects. (You should pass all the headers in its constructor)